### ISSUE TITLE
Think about supporting the GraphQL API

### ISSUE THREAD

**Issue by max-wittig** [python-gitlab/python-gitlab#1357]
Not sure, if this would be worth a separate library, but doesn't look like Gitlab will depreciate the v4 API anytime soon.

So just something to keep in mind.

https://docs.gitlab.com/ee/api/graphql/reference/index.html

**Comment by nejch** [comment]
That's funny last weekend I was hacking away locally on a branch to wrap https://github.com/graphql-python/gql with existing authentication/convenience stuff from `gitlab.Gitlab` 😁

IMO it would also make sense to provide a thin wrapper at least! Since `gql` has a DSL as well, not sure how really needs to be done on top of that.

**Comment by max-wittig** [comment]
haha nice 👍  I just saw a depreciation warning somewhere in the docs (removed from v5 API) and checked out their GraphQL stuff again. Last time I checked they only had a project list endpoint.

**Comment by irishismyname** [comment]
FWIW some of the existing REST APIs are unstable and marked for deprecation, notably [Vulnerabilities](https://docs.gitlab.com/ee/api/vulnerabilities.html) and [Vulnerability Findings](https://docs.gitlab.com/ee/api/vulnerability_findings.html). I'm pulling my hair out trying to develop a vulnerability report ATM because this python-gitlab doesn't support GraphQL (yet?) and the existing REST APIs are broken.

Also https://github.com/python-gitlab/python-gitlab/issues/2316#issuecomment-1874627875.

**Comment by TimKnight-DWP** [comment]
In addition to whay @irishismyname posted above, they are changing how the CI_JOB_TOKEN works in GL 17.* and look to only be adding API control of "Allowlists" via the GraphQL API: https://gitlab.com/gitlab-org/gitlab/-/issues/435903

**Comment by mitar** [comment]
I think it would be nice if this library abstracted out the REST vs. GraphQL difference and one could just use the library, not caring what is used internally.

**Comment by TimKnight-DWP** [comment]
100% the end user of this library shouldn't need to worry about if it's GQL or REST, solely that they can perform a given action

**Comment by irishismyname** [comment]
I started writing a GitLab specific wrapper for `gql` and I have to say it has been quite a challenge to maintain backwards compatibility with this package and the REST API. The data structures _can_ be the same if you write your query very carefully, but the field names are just slightly different (camel vs. underscore). There are some places (like vulnerability findings) where the data structure needs a bunch of logic to map the new fields back what REST would have returned because they took a very different approach to the structure (e.g., vulnerability location).

**Comment by mitar** [comment]
My take would be that the library should not try to use both APIs for the same logical endpoint, but that it should use GraphQL only for those which do not exist in REST or have very limited support in REST. Like new endpoints which are added only in GraphQL.

**Comment by TimKnight-DWP** [comment]
@mitar agreed, and as REST endpoints are deprecated those get swapped to GQL.

Specifically we'll want to add support for editign `ci_job_token` allowlists which I suspect will only be available on GQL in Gitlab 17: https://gitlab.com/gitlab-org/gitlab/-/issues/435903

**Comment by amai2012** [comment]
See [REST API: Runner will no longer return ip_address](https://gitlab.com/gitlab-org/gitlab/-/issues/415159) and https://docs.gitlab.com/ee/api/runners.html states
`The ip_address attribute in the response was deprecated [in GitLab 16.1](https://gitlab.com/gitlab-org/gitlab/-/issues/415159) and will be removed in [a future version of the REST API](https://gitlab.com/gitlab-org/gitlab/-/issues/351109). This attribute will start returning an empty string in GitLab 17.0. The ipAddress attribute can be found inside the respective runner manager, currently only available through the GraphQL [CiRunnerManager type](https://docs.gitlab.com/ee/api/graphql/reference/index.html#cirunnermanager).`

**Comment by samakshkhatri-crest** [comment]
Hi, wanted to check in and see if there is any support planned for GraphQL. There are differences in response data between both REST and GraphQL APIs. I had raised an [issue ](https://gitlab.com/gitlab-org/gitlab/-/issues/464260#note_1929053845 )in Gitlab core to address this inconsistency but it doesn't seem like they wanna address it

**Comment by dacevedo12** [comment]
This would enable https://github.com/python-gitlab/python-gitlab/issues/2550 by being able to call the `createWorkItem` mutation

**Comment by JohnVillalovos** [comment]
NOTE: An example of a project with a Python based GraphQL client. I think it is fairly rudimentary but I haven't looked in a lot of detail.

https://gitlab.com/cki-project/cki-lib/-/blob/7f7478b676aac97dd38b5cc6febafd0cb7604f6f/cki_lib/gitlab.py#L51-117

**Comment by dacevedo12** [comment]
Can confirm the low level http_post works fine

```python
gitlab.http_post(
    "https://gitlab.com/api/graphql",
    post_data={
        "operationName": "createWorkItem",
        "query": """
            mutation createWorkItem($input: WorkItemCreateInput!) {
                workItemCreate(input: $input) {
                    errors
                }
            }
        """,
        "variables": {}
        },
    },
)
```

**Comment by nejch** [comment]
Hey everyone!

We now have a very basic GraphQL client as of v4.11.1, built on top on gql. It's currently just a thin wrapper around gql, so you can reuse its variable injection and DSL if you like (see upstream docs). It really only handles sessions, authentication, retries and error handling so you can have a more robust library for GitLab specifically, the way you're used to from our REST client.

The next step will be to handle pagination in the client itself, but we just need to come up with an elegant way to do it. As you can see with the cki-lib linked above, there are so many ways to paginate at different levels in GraphQL, you might need to force an opinionated way of doing it to make it "generic". This is why gql itself also explicitly does not try to handle pagination AFAIK.

Regarding having an API-agnostic abstraction: based on my experience so far, the GraphQL and REST APIs are just entirely different beasts and not sure we can reliably do that here, especially as graphql can sometimes be more about efficient one-shot query building rather than writing procedural code. One thing we could probably do easily is make the data prettier by creating objects with snake_case attribute access rather than just plain dictionaries with JSON responses. If anyone has better ideas here please share!

gql does allow pre-fetching the qraphql schema or even using a local schema, and that might be interesting for validation and even maybe IDE autocompletion if we go wild, but I haven't really explored that yet. More of a follow-up issue.

**Comment by TheLastProject** [comment]
I've recently used the new GraphQL support to modify the CI_JOB_TOKEN CI/CD access programmatically and after taking some time to learn how GraphQL works I must say that it seems to work quite well! Thanks a lot for this support!

One thing I ran into while working with this is that I need to specify a custom CA location to talk to the GitLab instance we're using this with and noticed that the documentation on https://python-gitlab.readthedocs.io/en/stable/api-usage-advanced.html#ssl-certificate-verification does not work for the new GraphQL client. Digging into the change of #2967 a bit I quickly noticed that the new GraphQL support in python-gitlab uses HTTPX instead of Requests and thus needs other environment variables: https://www.python-httpx.org/compatibility/#ssl-configuration

This is not a huge deal, but I'm mostly wondering: is it intended to need to set `REQUESTS_CA_BUNDLE` for "regular" python-gitlab requests and `SSL_CERT_FILE` or `SSL_CERT_DIR` for python-gitlab GraphQL requests? I don't think this is a huge problem when documented but I was just curious if there was an explicit reason HTTPX was chosen for GraphQL support when Requests also seemed supported and would probably cause more consistency in environment variable usage for these cases: https://gql.readthedocs.io/en/stable/transports/requests.html

But yeah, aside from this minor documentation(?) issue which is quite understandable especially for a feature clearly marked as beta it is serving me quite well so far! :)

**Comment by nejch** [comment]
Thanks a lot for the feedback @TheLastProject!

This is somewhat expected yes, because we're essentially relying on the behavior of underlying libraries rather than explicit configuration. You should be able to also set this programmatically if you supply your own `Client` to our `GraphQL` instance. We should definitely document this a little better!

httpx was chosen on purpose as it allows us to add an async client easily, which was a big blocker for the REST 
client where we still rely on requests (see https://github.com/python-gitlab/python-gitlab/issues/1025, another popular request). 

Since they're mostly compatible, I think in the long run we might want to switch to httpx entirely but this would also be a big breaking change. It also feels like async-compatible libraries get a little more attention in `gql`.

**Comment by ArtemIsmagilov** [comment]
What do you think about this library that is asynchronously compatible and supports the Requests interface? Does it have any advantages over httpx?
- https://github.com/jawah/niquests

**Comment by nejch** [comment]
> What do you think about this library that is asynchronously compatible and supports the Requests interface? Does it have any advantages over httpx?
> 
>     * https://github.com/jawah/niquests

@ArtemIsmagilov I'm not sure it's worth switching to a third HTTP requests library at this point - httpx is also sync/async compatible and broadly behaves like requests as well. It'll actually be trivial to add an async GraphQL client using httpx now that we have basic support for it, I'll work on this in the coming weeks.

Also httpx is quite popular and maintained by the encode folks, i.e. the author of mkdocs, Django REST framework etc, so I think it's in good hands. If I missed something about the other library though just let me know!

**Comment by nejch** [comment]
The async part is coming with https://github.com/python-gitlab/python-gitlab/pull/3045, so additional feedback welcome if anyone takes it for a spin! :bow:
